tests/test_buffered_synced_collection.py

```diff
@@ -0,0 +1,421 @@
+            assert 'buffered' in synced_dict
+            assert synced_dict['buffered'] == testdata
+            synced_dict['buffered2'] = 1
+            assert 'buffered2' in synced_dict
+            assert synced_dict['buffered2'] == 1
+        assert len(synced_dict) == 2
+        assert 'buffered2' in synced_dict
+        assert synced_dict['buffered2'] == 1
+        with synced_dict.buffered():
+            del synced_dict['buffered']
+            assert len(synced_dict) == 1
+            assert 'buffered' not in synced_dict
+        assert len(synced_dict) == 1
+        assert 'buffered' not in synced_dict
+        assert 'buffered2' in synced_dict
+        assert synced_dict['buffered2'] == 1
+
+        # Explicitly check that the file has not been changed when buffering.
+        raw_dict = synced_dict.to_base()
+        with synced_dict.buffered():
+            synced_dict['buffered3'] = 1
+            on_disk_dict = self.load(synced_dict)
+            assert 'buffered3' not in on_disk_dict
+            assert on_disk_dict == raw_dict
+
+        on_disk_dict = self.load(synced_dict)
+        assert 'buffered3' in on_disk_dict
+        assert on_disk_dict == synced_dict
+
+    def test_two_buffered(self, synced_dict, testdata):
+        """Test that a non-buffered copy is not modified."""
+        synced_dict['buffered'] = testdata
+        synced_dict2 = BufferedJSONDict(filename=synced_dict._filename)
+
+        # Check that the non-buffered object is not modified.
+        with synced_dict.buffered():
+            synced_dict['buffered2'] = 1
+            assert 'buffered2' not in synced_dict2
+
+    def test_two_buffered_modify_unbuffered(self, synced_dict, testdata):
+        """Test that in-memory changes raise errors in buffered mode."""
+        synced_dict['buffered'] = testdata
+        synced_dict2 = BufferedJSONDict(filename=synced_dict._filename)
+
+        # Check that the non-buffered object is not modified.
+        with pytest.raises(MetadataError):
+            with synced_dict.buffered():
+                synced_dict['buffered2'] = 1
+                synced_dict2['buffered2'] = 2
+                assert synced_dict['buffered2'] == 1
+                synced_dict['buffered2'] = 3
+                assert synced_dict2['buffered2'] == 2
+                synced_dict2['buffered2'] = 3
+                assert synced_dict['buffered2'] == 3
+
+        self._force_flush_backends()
+
+    def _force_flush_backends(self):
+        # TODO: If client code catches errors raised due to invalid data in the
+        # buffer then attempts to continue, the buffer will be in an invalid
+        # state, and any future attempt to leave buffered mode will trigger
+        # another flush that will error. I'm not sure what the best way to deal
+        # with this problem is. The easiest option is probably to clear the
+        # cache whenever an error occurs so that the buffer is back in a valid
+        # state.
```

Comment:
You're getting the gist of the problem, but I don't think you're quite grasping the cases that I want to avoid. Let me provide an example:

For now, let's assume that we're only dealing with a simple in-memory dict as a cache (the current behavior in signac). I agree that Redis introduces additional complexities, which we can now ignore for the purposes of this pull request. Now assume that the user does something that leads to a `MetadataError`, but they put this code into a try-except block so that they can catch the error. If they try and make _any_ modification of that collection again after the exception has been raised, it will re-raise the same exception because the cache remains invalid (the metadata will always look out-of-date). Therefore, catching a `MetadataError` becomes basically an invalid code path because there's no real way to recover from it. At minimum, I think this needs to be documented and `MetadataError` should clearly state that recovery is not possible.

However, it would be _nice_ if we could provide some additional guarantees of safety. There are additional problems that arise in global buffering; if half the flushes work before one fails, then half the user's changes will have been persisted to disk while all of the remaining changes (of which potentially there's only one invalid one) won't have been. That might itself result in an invalid state for the set of collections; we wouldn't know what the "right" solution is, but a user could try to catch the `MetadataError` in an attempt to correct for that problem. Making our code robust to that wouldn't be too difficult; the buffer could retain an `original_contents` field so that we could always restore to the pre-buffered state for all changes that don't introduce metadata errors. Of course then we could run into race conditions if something else is indeed modifying the file while this process is running, but we make no promise to support concurrent access.

I'm not saying we have to address all of these edge cases, but the ones that aren't too difficult I would like to at least fail very loudly and clearly if we choose not to help the user.

Reply:
> At minimum, I think this needs to be documented and `MetadataError` should clearly state that recovery is not possible.

Yes, I think that is the right choice. I would document "Raises: `MetadataError`" for the context managers, explain that signac will not attempt to recover, and state that clearing will allow the user to return to a valid state at the cost of potential data loss. Then if a user catches this error on purpose, it should be up to the user to attempt to fix the problem (clearing the buffer) or to exit as gracefully as possible.

To improve safety in the case you mentioned, it might be possible to defer raising the `MetadataError` until the buffer writes out all the data that hasn't been modified in the interim?

The idea you suggest for `original_contents` might be useful but I would prefer to "not clear" and let the user handle the problem. And the buffer class and the global buffer data storage are private anyway, right?

Reply:
Right, that's the problem. The details of the cache are all internal, so the user has no way to clear the cache. _That effectively means that there is no way for a user to recover from this state even if they catch the error on purpose, the only option is to exit "as gracefully as possible", which might not be very graceful given this constraint._ Is that really the behavior that we want?

Reply:
Discussed with @vyasr. We could return all unwritten buffer contents as an attribute of the `BufferedError` and clear the buffer's remaining data. It might be possible to add a method to the `BufferedError` that would forcibly overwrite the data on disk with the unwritten buffer contents.

Reply:
@bdice and I chatted about this briefly. We're in agreement that we don't need to support rollback in buffered mode. The current code already returns the data in the buffer within the Exception, so it makes sense to go ahead and clear the buffer contents so that future operations will work again if the user chooses to catch the exception and continue.